applications/GeoMechanicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_linear_elastic_dynamic.h

```diff
@@ -0,0 +1,703 @@
+        }
+    }
+
+    void Build(typename TSchemeType::Pointer pScheme, ModelPart& rModelPart, TSystemMatrixType& rA, TSystemVectorType& rb) override
+    {
+        Timer::Start("Build");
+        this->BuildLHS(pScheme, rModelPart, rA);
+        this->BuildRHSNoDirichlet(pScheme, rModelPart, rb);
+
+        Timer::Stop("Build");
+
+        TSystemVectorType dummy_b(rA.size1(), 0.0);
+        TSystemVectorType dummy_rDx(rA.size1(), 0.0);
+
+        // apply constraints
+        if (rModelPart.MasterSlaveConstraints().size() != 0) {
+            const auto timer_constraints = BuiltinTimer();
+            Timer::Start("ApplyConstraints");
+            BaseType::ApplyConstraints(pScheme, rModelPart, rA, rb);
+            BaseType::ApplyConstraints(pScheme, rModelPart, mMassMatrix, dummy_b);
+            BaseType::ApplyConstraints(pScheme, rModelPart, mDampingMatrix, dummy_b);
+            Timer::Stop("ApplyConstraints");
+            KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverLinearElasticDynamic", BaseType::GetEchoLevel() >= 1)
+                << "Constraints build time: " << timer_constraints << std::endl;
+        }
+
+        // apply dirichlet conditions
+        BaseType::ApplyDirichletConditions(pScheme, rModelPart, rA, dummy_rDx, rb);
+        BaseType::ApplyDirichletConditions(pScheme, rModelPart, mMassMatrix, dummy_rDx, dummy_b);
+        BaseType::ApplyDirichletConditions(pScheme, rModelPart, mDampingMatrix, dummy_rDx, dummy_b);
+
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverLinearElasticDynamic", BaseType::GetEchoLevel() >= 3)
+            << "Before the solution of the system"
+            << "\nSystem Matrix = " << rA << "\nUnknowns vector = " << dummy_rDx
+            << "\nRHS vector = " << rb << std::endl;
+
+        if (mCalculateInitialSecondDerivative) {
+            this->CalculateInitialSecondDerivative(rModelPart, rA, rb, pScheme);
+            mCopyExternalForceVector = true;
+        }
+
+        // only add dynamics to lhs after calculating initial second derivative
+        this->AddDynamicsToLhs(rA, rModelPart);
+
+        // Initialize the linear solver, such that the solver can factorize the matrices already. In
+        // case the matrices can be pre-factorized, this step is not performed during calculation.
+        BaseType::mpLinearSystemSolver->InitializeSolutionStep(rA, dummy_rDx, rb);
+
+        // check if PerformSolutionStep can be performed instead of Solve, this is more efficient for solvers, which can be pre-factorized
+        try {
+            BaseType::mpLinearSystemSolver->PerformSolutionStep(rA, dummy_rDx, rb);
+            mUsePerformSolutionStep = true;
+        } catch (const Kratos::Exception& e) {
+            std::string error_message = e.what();
+
+            // if PerformSolutionStep is not implemented, the following error is thrown, in this case, use Solve
+            if (error_message.find("Error: Calling linear solver base class") != std::string::npos) {
```

Comment:
The error message can be moved into the if to limit the scope: 
```suggestion
            if (std::string error_message = e.what(); error_message.find("Error: Calling linear solver base class") != std::string::npos) {

```